doc/neps/nep-0021-advanced-indexing.rst

```diff
@@ -11,30 +12,175 @@ Implementing intuitive and full featured advanced indexing
+  new axis of size 1 at the end. Basic indexing always return a view of the
+  indexed array's data.
+- "Advanced" indexing, also called "fancy" indexing, includes all cases where
+  arrays are indexed by other arrays. Advanced indexing always makes a copy:
+
+  - "Boolean" indexing by boolean arrays, e.g., ``x[x > 0]`` for
+    selecting positive elements.
+  - "Vectorized" indexing by one or more integer arrays, e.g., ``x[[0, 1]]``
+    for selecting the first two elements along the first axis. With multiple
+    arrays, vectorized indexing uses broadcasting rules to combine indices along
+    multiple dimensions. This allows for producing a result of arbitrary shape
+    with arbitrary elements from the original arrays.
+  - "Mixed" indexing involving any combinations of the other advancing types.
+    This is no more powerful than vectorized indexing, but is sometimes more
+    convenient.
+
+For clarity, we will refer to these existing rules as "legacy indexing".
+This is only a high-level summary; for more details, see NumPy's documentation
+and and `Examples` below.
+
+Outer indexing
+~~~~~~~~~~~~~~
+
+One broadly useful class of indexing operations is not supported:
+
+- "Outer" or orthogonal indexing treats one-dimensional arrays equivalently to
+  slices for determining output shapes. The rule for outer indexing is that the
+  result should be equivalent to independently indexing along each dimension
+  with integer or boolean arrays as if both the indexed and indexing arrays
+  were one-dimensional. This form of indexing is familiar to many users of other
+  programming languages such as MATLAB, Fortran and R.
+
+The reason why NumPy omits support for outer indexing is that the rules for
+outer and vectorized conflict. Consider indexing a 2D array by two 1D integer
+arrays, e.g., ``x[[0, 1], [0, 1]]``:
+
+- Outer indexing is equivalent to combining multiple integer indices with
+  ``itertools.product()``. The result in this case is another 2D array with
+  all combinations of indexed elements, e.g.,
+  ``np.array([[x[0, 0], x[0, 1]], [x[1, 0], x[1, 1]]])``
+- Vectorized indexing is equivalent to combining multiple integer indices with
+  ``zip()``. The result in this case is a 1D array containing the diagonal
+  elements, e.g., ``np.array([x[0, 0], x[1, 1]])``.
+
+This difference is a frequent stumbling block for new NumPy users. The outer
+indexing model is easier to understand, and is a natural generalization of
+slicing rules. But NumPy instead chose to support vectorized indexing, because
+it is strictly more powerful.
+
+It is always possible to emulate outer indexing by vectorized indexing with
+the right indices. To make this easier, NumPy includes utility objects and
+functions such as ``np.ogrid`` and ``np.ix_``, e.g.,
+``x[np.ix_([0, 1], [0, 1])]``. However, there are no utilities for emulating
+fully general/mixed outer indexing, which could unambiguously allow for slices,
+integers, and 1D boolean and integer arrays.
```

Comment:
Are you suggesting we do not allow higher dim bools in `oindex`? It seems rather useful to me, even if basically just a helper around nonzero.

Reply:
The possibility of allowing booleans with more than one dimension in oindex never occurred to me. I will need to ponder it.

Reply:
I think I tried to add support for that to `np.ix_` at some point in the past..

Reply:
I think it definitely makes sense, unless you want to do something like allow a higher D boolean, but not in combination with others, which seems strange to me. It is also a nice thing about forcing the correct dimension for the index itself, I think that might reduce bugs for such cases slightly.

What I could imagine discussing is to not quite allow all such cases for the final state of plain indexing.

About `np.ix_` I think adding it shouldn't be too hard (with the exception of 0d bools maybe, would have to think about it), but if this moves ahead...